### PR TITLE
new post: multi-store - Custom Telemetry with shared data

### PR DESCRIPTION
**Pull request by badboy**
[Rendered](https://github.com/badboy/fnordig.de/pull/8/files?short_path=5b07bbd#diff-5b07bbd13303298ea7040456c84bbf04)

### CODE REVIEW

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+Some data should not be correlated with other data due to privacy concerns.
+So far we had to push Telemetry users to create custom pings and keep track of their own data.
+If they rely on scalars or histograms as recorded my Telemetry core, but send a custom ping in different intervals, they can't make valid assumptions about the metrics, as they might have been reset in between.
+This leads to weird hacks or unnecessary code duplication. Additionally we can't provide any help or support for custom data and our tools can't handle it automatically (e.g. to generate dashboards).
+
+### A solution
+
+**Multi-Store**.
+
+Every probe was always tied to the schedule of the "main" ping (read: subsession/session).
```
Comment (Dexterp37) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: Are we calling them probes or metrics? I think we settled on "metrics", at least on mobile.

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+Our **multi-store** solution now enables probes to be associated with multiple stores at once, defaulting to be in the "main" ping only.
+Any user with custom requirements can now select probes to be included in their custom store (which is still subject to [Data Collection Review](https://wiki.mozilla.org/Firefox/Data_Collection)).
+Telemetry Core is still responsible for actual data storage and the APIs, but now the custom ping is responsible for collecting, clearing and periodically sending this data.
```
Comment (Dexterp37) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: I'd say that "Telemetry" is enough, do we really need the "Core" part? To the outside, its just Telemetry.

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+### An example
+
+To demonstrate how this is used, let's create a custom ping, which will include one probe of its own and one that's generally available.
```
Comment (Dexterp37) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: "one metric of its own"?

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+We start by adding a new probe to [Histograms.json][] (somewhere in the file, I placed it just above `TELEMETRY_EVENT_PING_SENT`, but it doesn't matter much):
```
Comment (Dexterp37) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: Missing link? Also, is the one within `()` a note you forgot to cleanup?

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
++    "record_into_store": ["main", "custom-store"]
+   },
+   "GC_BUDGET_MS": {
+     "record_in_processes": ["main", "content"],
+```
+
+Next, we define our new custom ping in a new file in `toolkit/components/telemetry/pings/CustomPing.jsm`.
+We only define a very simple interface: A way to start the custom ping, which itself sets up a (persistent) timer (firing every 24 hours).
+When fired, `notify()` will be called, which then collects the payload from `custom-store` and schedule the ping for sending.
+Telemetry Core will take care of actually sending the ping (and retrying, storing it in the archive, etc.).
```
Comment (Dexterp37) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: Telemetry vs Telemetry Core?

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+That is right. Let's add some random values once in a while:
+
+```patch
+--- c/toolkit/components/telemetry/pings/TelemetrySession.jsm
++++ i/toolkit/components/telemetry/pings/TelemetrySession.jsm
+@@ -440,6 +440,10 @@ var TelemetryScheduler = {
+   _schedulerTickLogic() {
+     this._log.trace("_schedulerTickLogic");
+
++    // Adding a random value at every tick.
```
Comment (Dexterp37) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: I'd avoid to publish an example with a change to `TelemetrySession.jsm`. Why would you need this anyway? You're populating the histogram in the custom ping file.

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+```
+./mach run
+```
+
+Once Telemetry is initialized (should take 60 seconds), you should be able to see the freshly generated custom ping (in a development build this ping is never actually send out).
+To see it, go to `about:telemetry`, click on `Current Ping` in the upper-left corner, select `Archived ping data` and then the `custom` ping type. There you have it!
+
+![Raw payload ouf the custom ping](https://tmp.fnordig.de/blog/2019/about-telemetry-customping.png)
+
+The full changeset is available [in this commit on mozilla-central](https://hg.mozilla.org/try/rev/2d6a3ea139b3b8f23e5ccb328c952d726e9829ef) (don't worry, this is not landing in Firefox).
```
Comment (Dexterp37) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: Are try changesets living long enough?

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md (hunk 1 above)
Comment (badboy) [reply _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: Good call. I will unify it to use metrics.

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md (hunk 2 above)
Comment (badboy) [reply _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: Alright

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md (hunk 4 above)
Comment (badboy) [reply _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: The wonders of markdown. The link is defined further down, so this works just fine. The () sentence can go away indeed.

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md (hunk 6 above)
Comment (badboy) [reply _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: True. Removing it.

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md (hunk 7 above)
Comment (badboy) [reply _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: No idea at all. Gonna move this to GitHub.

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md (hunk 4 above)
Comment (Dexterp37) [reply _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: TIL!

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+Last year I implemented a new feature for Firefox Telemetry that changes how we can collect and analyze data with different requirements in regard to user privacy & frequency.
+This post will shine some light on the (rather simple) implementation and usage.
+
+### Intro: What is Firefox Telemetry?
+
+In order to understand how Firefox performs in the wild, it can collect a bunch of performance metrics and other information.
+The [Telemetry component](https://searchfox.org/mozilla-central/source/toolkit/components/telemetry/) provides APIs used by the various other parts to gather this data
+and is responsible for storing, collecting and sending this data in what we call "pings", a periodic collection of measurements.
+Telemetry data is only ever send out if the user agreed to it (see "Data Collection and Use" in the "Privacy & Security" preferences of your Firefox).
```
Comment (georgf) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: Nit: "sent"

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+Data can be collected in one of 3 different formats: histograms, scalars & events (see [our collection overview](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/collection/index.html)).
```
Comment (georgf) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: "Most data is collected in ..."?

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+Firefox sends this data in the ["main" ping](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/data/main-ping.html) once in a while (usually roughly daily) and clears out the stored data locally.
+A "main" ping always corresponds to a subsession, which itself is part of a session, this is further explained [in our Session concepts](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/concepts/sessions.html).
```
Comment (georgf) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: Nit: "... session. This is ..."

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+People working with the data can therefore make certain assumptions on how to interpret the data from multiple pings across sessions and subsessions.
+
+### The problem
+
+Some data should not be correlated with other data due to privacy concerns.
+So far we had to push Telemetry users to create custom pings and keep track of their own data.
+If they rely on scalars or histograms as recorded my Telemetry, but send a custom ping in different intervals, they can't make valid assumptions about the metrics, as they might have been reset in between.
```
Comment (georgf) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: "by Telemetry"

Comment (georgf) [reply _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: Does it make sense to add a simple example here?

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+Every metric was always tied to the schedule of the "main" ping (read: subsession/session).
+Our **multi-store** solution now enables metrics to be associated with multiple stores at once, defaulting to be in the "main" ping only.
+Any user with custom requirements can now select metrics to be included in their custom store (which is still subject to [Data Collection Review](https://wiki.mozilla.org/Firefox/Data_Collection)).
+Telemetry is still responsible for actual data storage and the APIs, but now the custom ping is responsible for collecting, clearing and periodically sending this data.
+
+### An example
+
+To demonstrate how this is used, let's create a custom ping, which will include one metric of its own and one that's also available in the "main" store (and thus the "main" ping).
+
+We start by adding a new metric to [Histograms.json](https://searchfox.org/mozilla-central/source/toolkit/components/telemetry/Histograms.json):
```
Comment (georgf) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: One totally optional thought:
Could we use a scalar as an example, to shift the attention a bit away from histograms?

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+data:
+  route: blog
+---
+
+Last year I implemented a new feature for Firefox Telemetry that changes how we can collect and analyze data with different requirements in regard to user privacy & frequency.
+This post will shine some light on the (rather simple) implementation and usage.
+
+### Intro: What is Firefox Telemetry?
+
+In order to understand how Firefox performs in the wild, it can collect a bunch of performance metrics and other information.
```
Comment (georgf) [inline _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: Could add a nice link here that clearly introduces people to our Firefox data practices?
A good blog post or wiki article or similar. Maybe Chris has a good link handy?

**Review by georgf — APPROVED** [review]

File: _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md (hunk 11 above)
Comment (badboy) [reply _posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md]: I can't come up with a simple one. I don't think it's necessary to provide it here.

**Pull request merged**